### PR TITLE
[Inference] Tensor model parallelism

### PR DESCRIPTION
**Pull request by gabrieleoliaro**
**Description of changes:**

**Related Issues:**

Linked Issues:
- Issue #

Issues closed by this PR:
- Closes #

**Before merging:**

- [ ] Did you update the [flexflow-third-party](https://github.com/flexflow/flexflow-third-party) repo, if modifying any of the Cmake files, the build configs, or the submodules?

### CODE REVIEW

**Review by jiazhihao — APPROVED** [review]

**Comment by jiazhihao** [comment]
@gabrieleoliaro The PR looks great to me overall. Do you think we can merge this into the inference branch?

File: src/ops/spec_inc_multihead_self_attention.cc [hunk 1]
```diff
@@ -127,7 +127,7 @@ Tensor
   }
   if (bias) {
     // q, k, v, o
-    int dims[1] = {embed_dim * 4};
+    int dims[1] = {(qProjSize + kProjSize + vProjSize) * num_heads + oProjSize};
```
Comment (xinhaoc) [inline src/ops/spec_inc_multihead_self_attention.cc:130]: Hi Gabriele, just to make sure, do we have any situation that the `embed_dim * 4` is not equal to `qProjSize + kProjSize + vProjSize) * num_heads + oProjSize`?

Comment (jiazhihao) [reply src/ops/spec_inc_multihead_self_attention.cc:130]: In Megatron-LM's tensor model parallelism, we partition attention heads across GPUs, so `num_heads` of each GPU 
can be smaller than the total number of attention heads

Comment (gabrieleoliaro) [reply src/ops/spec_inc_multihead_self_attention.cc:130]: In the current model we have, the two quantities are indeed always identical. I changed this just to make it easier to remember the data layout

**Comment by gabrieleoliaro** [comment]
> @gabrieleoliaro The PR looks great to me overall. Do you think we can merge this into the inference branch?

yes! I'll merge it as soon as CI passes!

File: src/ops/spec_inc_multihead_self_attention.cc (hunk 1 above)
Comment (xinhaoc) [reply src/ops/spec_inc_multihead_self_attention.cc:130]: thanks!

**Pull request merged**